File: OI/jzoj/1262.pas
var	i,j,k,n,l,t,ans,tmp:longint;
	s:string;
	st,m:int64;
procedure dfs(dep,cnt:longint);
var	i:longint;
begin
	if cnt<0 then exit;
	if dep>l-t then begin
		for i:=dep to l-1 do 
			if st and (1<<i)<>0 then dec(cnt);
		if (cnt=0) and (tmp<ans) then ans:=tmp;
		exit;
	end;
	for i:=0 to 1 do begin
		if i=0 then begin
			inc(tmp);
			st:=st xor (m<<dep);
			if st and (1<<dep)<>0 then 
				dfs(dep+1,cnt-1)
			else dfs(dep+1,cnt);
			st:=st xor (m<<dep);
			dec(tmp);
		end else 
			if st and (1<<dep)<>0 then 
				dfs(dep+1,cnt-1)
			else dfs(dep+1,cnt);
	end;
end;
begin
	assign(input,'xlite.in');reset(input);
	assign(output,'xlite.out');rewrite(output);

	st:=0;m:=0;
	readln(l,t);
	readln(s);
	for i:=1 to l do st:=st<<1+ord(s[i])-48;
	readln(s);
	for i:=1 to t do m:=m<<1+ord(s[i])-48;
	
	for i:=0 to l do begin
		ans:=maxlongint;
		dfs(0,i);
		if ans<>maxlongint then begin
			writeln(ans);
			halt;
		end;
	end;

	close(input);close(output);

end.

File: OI/jzoj/1262.cpp
#include <cstdio>

long long st=0,m=0,tt;
int l,t,ans,tmp=0;
char s[100];
int a[100];

void dfs(int dep,int cnt)
{
	if (cnt<0) return;
	if (dep>l-t) {
		for (int i=dep;i<l;i++){
			if ((st & (long long) 1<<i)!=0) cnt --;
		}
		if ((cnt==0) && (tmp<ans)) {
			ans=tmp;
		}
		return;
	}
	for (int i=0;i<=1;i++)
	{
		if (i==0) {
			a[dep]=true;
			tmp ++;
			st^=(long long) m<<dep;
			
			if ((st & (long long) 1<<dep)!=0){
				dfs(dep+1,cnt-1);
			}else {
				dfs(dep+1,cnt);
			}
			st^=(long long) m<<dep;
			tmp --;
			a[dep]=false;
		}else {

			if ((st & (long long) 1<<dep)!=0){
				dfs(dep+1,cnt-1);
			}else {
				dfs(dep+1,cnt);
			}
		}
	}
}

int main()
{
	freopen("xlite.in","r",stdin);
	freopen("xlite.out","w",stdout);
	
	scanf("%d%d\n",&l,&t);
	scanf("%s",&s);
	for (int i = 0;i < l;i ++) {
		st=(st<<1)+s[i]-'0';
	}
	scanf("%s",&s);
	for (int i = 0;i < t;i ++) m=(m<<1)+s[i]-48;

	for (int i=0;i<=l;i ++)
	{
		ans=1<<30;
		dfs(0,i);
		if (ans!=1<<30) 
		{
			printf("%d\n",ans);
			return 0;
		}
	}

	return 0;
}
